src/quick-lint-js/fe/parse-type.cpp

```diff
@@ -390,7 +391,21 @@ void parser::parse_and_visit_typescript_type_expression(parse_visitor_base &v) {
 void parser::parse_and_visit_typescript_colon_type_expression_or_type_predicate(
     parse_visitor_base &v) {
   this->parse_typescript_colon_for_type();
+  //: ?Type, invalid
+  if (this->peek().type == token_type::question) {
```

Comment:
[parse_and_visit_typescript_colon_type_expression_or_type_predicate](https://github.com/quick-lint/quick-lint-js/blob/master/src/quick-lint-js/fe/parse-expression.cpp#L1963) is called here for that statement and seems to be needed. Replacing it with parse_and_visit_typescript_colon_type_expression fails to parse some predicates.

Since we want to catch the error in any type expression and not just colon type expression, wouldn't it make sense to put the logic in [parser::parse_and_visit_typescript_type_expression](https://github.com/quick-lint/quick-lint-js/blob/master/src/quick-lint-js/fe/parse-type.cpp#L44)? I'm working on a fix there but I need to figure out how to make sure we're not in a tuple or after an _extends_.

Reply:
> Replacing it with parse_and_visit_typescript_colon_type_expression fails to parse some predicates.

Yes, I understand that. The code you linked is used of return types *and* parameter types; it doesn't know if it's a parameter or return type.

> Since we want to catch the error in any type expression and not just colon type expression, wouldn't it make sense to put the logic in [parser::parse_and_visit_typescript_type_expression](https://github.com/quick-lint/quick-lint-js/blob/master/src/quick-lint-js/fe/parse-type.cpp#L44)?

Yes.

> I need to figure out how to make sure we're not in a tuple or after an *extends*.

Hmm, yeah, that gets tricky. We could maybe add a boolean parameter (`parse_question_as_invalid`).